data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/ListToMapProcessor.java

```diff
@@ -0,0 +1,144 @@
+                LOG.error(EVENT, "Given source path [{}] is not valid on record [{}]",
+                        config.getSource(), recordEvent, e);
+                continue;
+            }
+
+            ObjectNode targetNode = null;
+            try {
+                targetNode = constructTargetNode(sourceNode);
+            } catch (IllegalArgumentException e) {
+                LOG.error(EVENT, "Cannot find a list at the given source path [{}} on record [{}]",
+                        config.getSource(), recordEvent, e);
+                continue;
+            } catch (final Exception e) {
+                LOG.error(EVENT, "Error converting source list to map on record [{}]", recordEvent, e);
+            }
+
+            try {
+                updateEvent(recordEvent, targetNode);
+            } catch (final Exception e) {
+                LOG.error(EVENT, "Error updating record [{}] after converting source list to map", recordEvent, e);
+            }
+        }
+        return records;
+    }
+
+    private JsonNode getSourceNode(final Event recordEvent) {
+        final Object sourceObject = recordEvent.get(config.getSource(), Object.class);
+        return OBJECT_MAPPER.convertValue(sourceObject, JsonNode.class);
+    }
+
+    private ObjectNode constructTargetNode(JsonNode sourceNode) {
+        final ObjectNode targetNode = OBJECT_MAPPER.createObjectNode();
+        if (sourceNode.isArray()) {
+            for (final JsonNode itemNode : sourceNode) {
+                String itemKey = itemNode.get(config.getKey()).asText();
+
+                if (!config.getFlatten()) {
+                    final ArrayNode itemValueNode;
+                    if (!targetNode.has(itemKey)) {
+                        itemValueNode = OBJECT_MAPPER.createArrayNode();
+                        targetNode.set(itemKey, itemValueNode);
+                    } else {
+                        itemValueNode = (ArrayNode) targetNode.get(itemKey);
+                    }
+
+                    if (config.getValueKey() == null) {
+                        itemValueNode.add(itemNode);
+                    } else {
+                        itemValueNode.add(itemNode.get(config.getValueKey()));
+                    }
+                } else {
+                    if (!targetNode.has(itemKey) || config.getFlattenedElement() == ListToMapProcessorConfig.FlattenedElement.LAST) {
+                        if (config.getValueKey() == null) {
+                            targetNode.set(itemKey, itemNode);
+                        } else  {
+                            targetNode.set(itemKey, itemNode.get(config.getValueKey()));
+                        }
+                    }
+                }
+            }
+        } else {
+            throw new IllegalArgumentException("Cannot find a list at the given source path [{}}" + config.getSource());
+        }
+        return targetNode;
+    }
+
+    private void updateEvent(Event recordEvent, ObjectNode targetNode) {
```

Comment:
Again, I think this could probably use `Map` instead of `ObjectNode`. This is not a necessary change since it is all internal to the code.